server/src-lib/Hasura/GraphQL/Resolve/Introspect.hs

```diff
@@ -230,22 +230,39 @@ namedTypeR nt fld = do
 
 namedTypeR'
   :: ( MonadReader r m, Has TypeMap r
-     , MonadError QErr m)
+     , MonadError QErr m, MonadReusability m)
   => Field
   -> TypeInfo
   -> m J.Object
-namedTypeR' fld = \case
-  TIScalar colTy        -> scalarR colTy fld
-  TIObj objTyInfo       -> objectTypeR objTyInfo fld
-  TIEnum enumTypeInfo   -> enumTypeR enumTypeInfo fld
-  TIInpObj inpObjTyInfo -> inputObjR inpObjTyInfo fld
-  TIIFace iFaceTyInfo   -> ifaceR' iFaceTyInfo fld
-  TIUnion unionTyInfo   -> unionR unionTyInfo fld
+namedTypeR' fld tyInfo = do
+  -- Here we make sure to read the 'includeDeprecated' argument to the
+  -- 'fields' and 'enumValues' fields, to ensure that the query is
+  -- marked unreusable if necessary (fix #4547).
+  _ <- forM (toList (_fSelSet fld)) $ \subFld ->
+    case _fName subFld of
+      "fields"     -> readIncludeDeprecated subFld
+      "enumValues" -> readIncludeDeprecated subFld
+      _            -> return ()
+  -- Now fetch the required type information from the corresponding
+  -- information generator
```

Comment:
We can parse the arguments for these fields in their respective type resolvers? In [`enumTypeR`](https://github.com/hasura/graphql-engine/pull/4661/files#diff-9539cee3bf7b3e5530f83d51b5ae1a0dL170) for example? And similarly in `objectTypeR` and `ifaceR'`?

Reply:
We also don't seem to use these arguments to generate the response as required by the spec. In case of enums it is trivial to fix. In case of objects and interfaces, currently we don't seem to track whether a field is deprecated or not. So we might as well fix that.

Reply:
I parse these arguments here, rather than in the individual fields, because the code would be the same in each of the fields.

> In case of objects and interfaces, currently we don't seem to track whether a field is deprecated or not. So we might as well fix that.

Where can we get deprecation information from?

Reply:
> I parse these arguments here, rather than in the individual fields, because the code would be the same in each of the fields.

Aren't we better off just writing it in the respective type handler? For example we seem to be reusing `objectTypeR` in `unionR`. 

> Do we currently store deprecation information anywhere?

`EnumValInfo` stores deprecation information in the `_eviIsDeprecated` field. This information seems to be missing in `ObjFldInfo`.

Reply:
If we were to fix handling `includeDeprecated` argument correctly, the deprecation information in case of fields for object types introduced through remote schemas has to be propagated from the introspection query.